crates/ruff/src/rules/pyupgrade/rules/use_pep695_type_alias.rs

```diff
@@ -75,15 +80,78 @@ pub(crate) fn non_pep695_type_alias(checker: &mut Checker, stmt: &StmtAnnAssign)
     //              as type params instead
     let mut diagnostic = Diagnostic::new(NonPEP695TypeAlias { name: name.clone() }, stmt.range());
     if checker.patch(diagnostic.kind.rule()) {
+        let mut visitor = TypeVarNameVisitor {
+            names: vec![],
+            semantic: checker.semantic(),
+        };
+        visitor.visit_expr(value);
+
+        let type_params = if visitor.names.is_empty() {
+            None
+        } else {
+            Some(ruff_python_ast::TypeParams {
+                range: TextRange::default(),
+                type_params: visitor
+                    .names
+                    .iter()
+                    .map(|name| {
+                        TypeParam::TypeVar(TypeParamTypeVar {
+                            range: TextRange::default(),
+                            name: Identifier::new(name.id.clone(), TextRange::default()),
+                            bound: None,
+                        })
+                    })
+                    .collect(),
+            })
+        };
+
         diagnostic.set_fix(Fix::automatic(Edit::range_replacement(
             checker.generator().stmt(&Stmt::from(StmtTypeAlias {
                 range: TextRange::default(),
                 name: target.clone(),
-                type_params: None,
+                type_params,
                 value: value.clone(),
             })),
             stmt.range(),
         )));
     }
     checker.diagnostics.push(diagnostic);
 }
+
+struct TypeVarNameVisitor<'a> {
+    names: Vec<&'a ExprName>,
+    semantic: &'a SemanticModel<'a>,
+}
+
+impl<'a> Visitor<'a> for TypeVarNameVisitor<'a> {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match expr {
+            Expr::Name(name) if name.ctx.is_load() => {
+                let Some(Some(Stmt::Assign(StmtAssign { value, .. }))) =
+                    self.semantic
+                        .scope()
+                        .get(name.id.as_str())
+                        .map(|binding_id| {
```

Comment:
Nit: You can use `and_then` to avoid `Option<Option<...>``
```suggestion
                        .and_then(|binding_id| {
```

Reply:
Ah thanks that was awkward